src/prepared_statement.c

```diff
@@ -0,0 +1,339 @@
+	if (found) {
+		/* Statement already prepared on this link, do not forward packet */
+		slog_debug(client, "handle_parse_command: mapping statement '%s' to '%s' (query '%s')", ps->name, link_ps->name, pp->query);
+
+		buf = create_parse_complete_packet();
+		if (!pktbuf_send_immediate(buf, client)) {
+			pktbuf_free(buf);
+			return false;
+		}
+
+		pktbuf_free(buf);
+
+		/* Register statement on client link */
+		HASH_ADD_KEYPTR(hh, client->prepared_statements, ps->name, strlen(ps->name), ps);
+	} else {
+		/* Statement not prepared on this link, sent modified P packet */
+		slog_debug(client, "handle_parse_command: creating mapping for statement '%s' to '%s' (query '%s')", ps->name, link_ps->name, pp->query);
+
+		buf = create_parse_packet(link_ps->name, ps->pkt);
+
+		/* update stats */
+		client->pool->stats.ps_server_parse_count++;
+
+		/* Track Parse command sent to server */
+		opp = calloc(sizeof *opp, 1);
+		opp->ignore = false;
+		list_append(&server->server_outstanding_parse_packets, &opp->node);
+
+		/* Register statement on client and server link */
+		HASH_ADD_KEYPTR(hh, client->prepared_statements, ps->name, strlen(ps->name), ps);
+		if (!register_prepared_statement(server, link_ps))
+			return false;
+
+		if (!pktbuf_send_queued(buf, server)) {
+			return false;
+		}
+	}
+
+	return true;
+}
+
+bool handle_bind_command(PgSocket *client, PktHdr *pkt)
+{
+	PgSocket *server = client->link;
+	struct PgBindPacket *bp;
+	PgParsedPreparedStatement *ps = NULL;
+	PgServerPreparedStatement *link_ps = NULL;
+	PktBuf *buf;
+	struct OutstandingParsePacket *opp = NULL;
+	bool found;
+	int diff;
+	uint32_t len;
+
+	Assert(server);
+
+	if (!unmarshall_bind_packet(client, pkt, &bp))
+		return false;
+
+	/* update stats */
+	client->pool->stats.ps_bind_count++;
+
+	HASH_FIND_STR(client->prepared_statements, bp->name, ps);
+	if (!ps) {
+		disconnect_client(client, true, "prepared statement '%s' not found", bp->name);
+		return false;
+	}
+
+	link_ps = get_prepared_statement(server, ps, &found);
+	if (!found) {
+		/* Statement is not prepared on this link, sent P packet first */
+		slog_debug(server, "handle_bind_command: prepared statement '%s' (query '%s') not available on server, preparing '%s' before bind", ps->name, ps->pkt->query, link_ps->name);
+
+		buf = create_parse_packet(link_ps->name, ps->pkt);
+
+		/* update stats */
+		client->pool->stats.ps_server_parse_count++;
+
+		/* Track Parse command sent to server */
+		opp = calloc(sizeof *opp, 1);
+		opp->ignore = true;
+		list_append(&server->server_outstanding_parse_packets, &opp->node);
+		if (!pktbuf_send_queued(buf, server)) {
+			return false;
+		}
+
+		/* Register statement on server link */
+		if (!register_prepared_statement(server, link_ps))
+			return false;
+	}
+	else if (HASH_COUNT(server->server_prepared_statements) != 1) /* do nothing if there is just one prepared statement */
+	{
+		/* Maintain LRU double-linke list: move element to the end of the list */
```

Comment:
_I'll create a thread here to keep the cache eviction mechanism discussion focussed in one place._

Regarding the cache algorithm, I agree there's many trade-offs here and supporting different eviction mechanisms would be nice. But I believe for the initial version of this prepared statement feature LRU is definitely good enough. The main case where LRU caches are problematic are when the full size of the cache is scanned, thus completely evicting the current set. This is especially problematic if this happens often.

I think it's pretty unlikely for that to happen for a decently sized (>=100 items) prepared statement cache. 

I agree with @knizhnik that LRU is at least better than least-often-used one, because in practice that will never evict existing items.

Reply:
To be clear. I think it would be beneficial to support multiple cache evection algorithms, but I couldn't quickly find a C implementation of W-TinyLFU. Having an implementation of that in PgBouncer/libusual might be desirable if LRU really has some problems for people in real world scenarios. But requiring that for this initial PR seems very unnecessary and will only make this PR harder to review.